presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TableLayoutRewriter.java

```diff
@@ -50,16 +52,18 @@
     private final Session session;
     private final SymbolAllocator symbolAllocator;
     private final PlanNodeIdAllocator idAllocator;
+    private final SqlParser sqlParser;
 
-    public TableLayoutRewriter(Metadata metadata, Session session, SymbolAllocator symbolAllocator, PlanNodeIdAllocator idAllocator)
+    public TableLayoutRewriter(Metadata metadata, Session session, SymbolAllocator symbolAllocator, PlanNodeIdAllocator idAllocator, SqlParser sqlParser)
     {
         this.metadata = metadata;
         this.session = session;
         this.symbolAllocator = symbolAllocator;
         this.idAllocator = idAllocator;
+        this.sqlParser = sqlParser;
     }
 
-    public PlanNode planTableScan(TableScanNode node, Expression predicate)
+    public Optional<PlanNode> planTableScan(TableScanNode node, Expression predicate)
```

Comment:
I don't think it should be `TableLayoutRewriter` responsibility to check that the plan has changed. It should be the rule that checks it since it is required in order to not loop forever. Other callers of `planTableScan` don't care if plan changed.